Examples/Io/Root/src/RootSimHitReader.cpp

```diff
@@ -0,0 +1,168 @@
+
+  if (m_cfg.filePath.empty()) {
+    throw std::invalid_argument("Missing input filename");
+  }
+  if (m_cfg.treeName.empty()) {
+    throw std::invalid_argument("Missing tree name");
+  }
+
+  m_outputSimHits.initialize(m_cfg.simHitCollection);
+
+  // Set the branches
+  int f = 0;
+  auto setBranches = [&](const auto& keys, auto& columns) {
+    for (auto key : keys) {
+      columns.insert({key, f++});
+    }
+    for (auto key : keys) {
+      m_inputChain->SetBranchAddress(key, &columns.at(key));
+    }
+  };
+
+  setBranches(m_floatKeys, m_floatColumns);
+  setBranches(m_uint32Keys, m_uint32Columns);
+  setBranches(m_uint64Keys, m_uint64Columns);
+  setBranches(m_int32Keys, m_int32Columns);
+
+  // add file to the input chain
+  m_inputChain->Add(m_cfg.filePath.c_str());
+  m_inputChain->LoadTree(0);
+  ACTS_DEBUG("Adding File " << m_cfg.filePath << " to tree '" << m_cfg.treeName
+                            << "'.");
+
+  // Because each hit is stored in a single entry in the root file, we need to
+  // scan the file first for the positions of the events in the file in order to
+  // efficiently read the events later on.
+  // TODO change the file format to store one event per entry
+
+  // Disable all branches and only enable event-id for a first scan of the file
+  m_inputChain->SetBranchStatus("*", false);
+  m_inputChain->SetBranchStatus("event_id", true);
+
+  auto nEntries = static_cast<std::size_t>(m_inputChain->GetEntriesFast());
+
+  // Add the first entry
+  m_inputChain->GetEntry(0);
+  m_eventMap.push_back({m_uint32Columns.at("event_id"), 0ul, 0ul});
+
+  // Go through all entries and store the position of new events
+  for (auto i = 1ul; i < nEntries; ++i) {
+    m_inputChain->GetEntry(i);
+    const auto evtId = m_uint32Columns.at("event_id");
+
+    if (evtId != std::get<0>(m_eventMap.back())) {
+      std::get<2>(m_eventMap.back()) = i;
+      m_eventMap.push_back({evtId, i, 0ul});
+    }
+  }
+
+  std::get<2>(m_eventMap.back()) = nEntries;
+
+  // Sort by event id
+  std::sort(m_eventMap.begin(), m_eventMap.end(),
+            [](const auto& a, const auto& b) {
+              return std::get<0>(a) < std::get<0>(b);
+            });
+
+  // Re-Enable all branches
+  m_inputChain->SetBranchStatus("*", true);
+}
+
+std::pair<size_t, size_t> ActsExamples::RootSimHitReader::availableEvents()
+    const {
+  return {std::get<0>(m_eventMap.front()), std::get<0>(m_eventMap.back())};
+}
+
+ActsExamples::ProcessCode ActsExamples::RootSimHitReader::read(
+    const ActsExamples::AlgorithmContext& context) {
+  auto it = std::find_if(
+      m_eventMap.begin(), m_eventMap.end(),
+      [&](const auto& a) { return std::get<0>(a) == context.eventNumber; });
+
+  if (m_inputChain == nullptr || it == m_eventMap.end()) {
+    ACTS_ERROR("Cannot read hits of event " << context.eventNumber);
+    return ActsExamples::ProcessCode::ABORT;
+  }
+
+  // lock the mutex
+  std::lock_guard<std::mutex> lock(m_read_mutex);
```

Comment:
does this make a difference for reading access?

Reply:
Hmm I don't think we can do this in parallel... I think we could if each thread would open its own `TFile` for reading.
But with this construction, we would have parallel modifying access to the "cursor" in the `TFile`, which probably leads to some kind of disaster...